game-core/src/main/java/games/strategy/triplea/util/MovableUnitsFilter.java

```diff
@@ -0,0 +1,198 @@
+package games.strategy.triplea.util;
+
+import com.google.common.base.Preconditions;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.MoveDescription;
+import games.strategy.engine.data.PlayerId;
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitCollection;
+import games.strategy.triplea.delegate.AbstractMoveDelegate;
+import games.strategy.triplea.delegate.AbstractMoveDelegate.MoveType;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.MoveValidator;
+import games.strategy.triplea.delegate.UndoableMove;
+import games.strategy.triplea.delegate.UnitComparator;
+import games.strategy.triplea.delegate.data.MoveValidationResult;
+import games.strategy.triplea.delegate.data.MustMoveWithDetails;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Utility class for filtering a unit list to the subset of those units that can move on a route.
+ */
+public final class MovableUnitsFilter {
+  /** The result of the filter operation. */
+  @AllArgsConstructor
+  @Getter
+  public static class Result {
+    // The filtered units and dependents that can move on the route.
+    private final List<Unit> unitsWithDependents;
+    // An error message, if no units could move on the route.
+    @Nullable private final String errorMessage;
+    // A warning message, if some, but not all units could move on the route.
+    @Nullable private final String warningMessage;
+  }
+
+  @AllArgsConstructor
+  @Getter
+  private static class MoveValidationResultWithDependents {
+    private final MoveValidationResult result;
+    private final List<Unit> unitsWithDependents;
+  }
+
+  private final GameData data;
+  private final PlayerId player;
+  private final Route route;
+  private final boolean nonCombat;
+  private final MoveType moveType;
+  private final List<UndoableMove> undoableMoves;
+
+  public MovableUnitsFilter(
+      final GameData data,
+      final PlayerId player,
+      final Route route,
+      final boolean nonCombat,
+      final MoveType moveType,
+      final List<UndoableMove> undoableMoves) {
+    this.data = Preconditions.checkNotNull(data);
+    this.player = Preconditions.checkNotNull(player);
+    this.route = Preconditions.checkNotNull(route);
+    this.nonCombat = nonCombat;
+    this.moveType = moveType;
+    this.undoableMoves = Preconditions.checkNotNull(undoableMoves);
+  }
+
+  /**
+   * Filters the units and their dependents to a subset of units that can move on the route.
+   *
+   * @param units The units to filter.
+   * @param dependentUnits The dependent units map.
+   * @return The result.
+   */
+  public Result filterUnitsThatCanMove(
```

Comment:
IIRC the agreement for code cov was to comment on warnings on why they were not appropriate to be fixed. Given we've extracted code to a new class, I would think it plausible to avoid code climate warnings. Is there strong reason to allow this warning? Ditto for the other code-climate warnings.

Reply:
@DanVanAtta suggestion on what can be simplified here?

Reply:
This specific comment by the way was on a previous version of the code before I moved some functionality to helpers. Now the function is ~30 lines of code.

One thing that feels like it would make sense to move is the logic inside the  ```!allUnitsResult.isMoveValid()``` block, but actually doing this exercise reveals that a lot of params have to be passed: lastUnitsResult, dependentUnits, route, best, etc. Which ends up being pretty complex.

I could make some of those be instance variables of the class. Right now, it's a bit arbitrary which parameters are passed to the ctor and which are passed to this function - so we could move everything to the ctor, which could simplify some of the internal passing of parameters between functions at the cost of keeping more state in the class. It's not obvious that's a win over just having the logic here.

By the way, the function is already significantly simplified over the original. Perhaps would should agree that "strictly better" changes are OK even if they trigger codeclimate warnings? @DanVanAtta WDYT?

Reply:
I now made a helper getInitialUnitList() which should make codeclimate happier. Personally, I don't find this actually helps readability much - because now it's less obvious what's going on - and the flow was already sequential. But it reduces lines of code here and complexity, so I guess it makes the metrics happy...

Reply:
@asvitkine you're right that we should not just make the metrics happy and seek to fundamentally improve the code. I think solving the code climate issues largely requires good faith efforts to spend more time resolving the issues. I recently had a method too large issue and eventually found a good way to shorten it up. My first reaction to the warning was negative though and thought the method seemed 'fine'.

IMO it's fine to respond to some of the warnings with "existing warning, this is moved code", and keep it short. I think we can all agree it's better if we can get these issues fixed, otherwise I wonder when if ever we'd fix them.

Reply:
Explicitly, if our goal to fix method too long warnings is to shorten 2 or 3 lines, then I'd agree we're only making the metrics happy. I suspect a method that is 'that' long should be broken up a good deal. If we just extract one or two lines to a helper method, the level of abstraction could become 'mixed'. EG: a long method with only low level code, if we extract just a bit to a method call, the method call will be a 'higher level' of abstraction, in which case we lots of low level code and then a method call. Hence, extracting more/most of the method into helper function is probably a good way to go.